Firefox 60esr

Mno-hime opened the pull request:
I put work by Geoffrey Weiss on top of previous @alarcher's branch, and then updated to 60.4.0esr and did couple of tweaks so it publishes. Likely ffmpeg needs to be added to build dependencies as well. p5m files needs to be reviewed.

I can see warnings like this:
```
[Child 2036, MediaPlayback #1] WARNING: Decoder=52c0e70 Decode error: NS_ERROR_DOM_MEDIA_FATAL_ERR (0x806e0005) - RefPtr<mozilla::MozPromise<RefPtr<mozilla::MediaTrackDemuxer::SamplesHolder>, mozilla::MediaResult, true> > mozilla::MediaSourceTrackDemuxer::DoGetSamples(int32_t): manager is detached.: file /export/home/newman/ws/oi-userland/components/web/firefox/firefox-60.4.0/dom/media/MediaDecoderStateMachine.cpp, line 3445
```
But no visible problem is associated with it.

However, when I synced by Firefox profile (setting, addons, ...), Firefox stuck and every couple of seconds printed:
```
[Parent 2002, Gecko_IOThread] WARNING: Failed to launch tab subprocess: file /export/home/newman/ws/oi-userland/components/web/firefox/firefox-60.4.0/ipc/glue/GeckoChildProcessHost.cpp, line 557
```

This is what Geoffrey has to say in email with the contribution:
```
With limited testing, these are problems:
* Encumbered HTML5 video/audio codecs do not work (the ffmpeg library
 autodetect not work with 64-bit?)
* In the window that shows certificate details (Page Info -> Security ->
 View Certificate), the tabs/notebook does not appear. This is probably a
 nimbus theme problem.
...
Things not tested: upgrading existing profile, wasm, web extensions,
editing userchrome.css, and most of the new technology added after FF52.
```

alarcher committed:
Update Firefox to 60esr

iigs commented:
64-bit Firefox won't work with 32-bit ffmpeg; it needs 64-bit ffmpeg.  Fix dependency issues if ffmpeg with 64-bit support doesn't install with a "pkg update".  No changes needed for building firefox.

If the gtk3 nimbus/B00merang theme is brought over to another OS, the tabs/notebook is missing in the View Certificate window.  If a popular theme like Adwaita is used, then the tabs/notebook is themed correctly on the View Certificate window.

HTML5 videos stutter, such as those on YouTube.  I have confirmed that at least VPx videos also stutter with Firefox 60.4ESR on Solaris 11.4.  This is a serious regression from 52ESR.

Upgrading existing profile worked on one system tested.  Web Assembly demos work.  Web Extensions work.  Customizing userchrome.css works.  Enabled technologies added after 52 work.

On an AMD Turion system, firefox can crash as it starts up, or after visiting a page or two.  The problem seen is described on this page: http://forums.mozillazine.org/viewtopic.php?f=38&t=3044432 .  The solution on that page is Linux specific.  This is a regression since 52ESR does not have this problem.  This is a ten year old system and is not a blocker as far as I am concerned.

After using firefox 60.4 for "a while", the firefox process has hung.  The window can not even be closed using the GUI/desktop controls. Killing the process from the command line stops it.  This behavior only seen on one system, but this system has not shown signs of instability in the past.

Show Memory report in about:memory does not work completely because jemalloc is disabled.  Compiling with jemalloc enabled results in a crash when running firefox.

Mno-hime commented:
@iigs How one enables ffmpeg in Firefox build? I added `video/ffmpeg` to my build zone and GZ where Firefox is run, I also added `CONFIGURE_OPTIONS +=   --enable-ffmpeg` to the Makefile, but after re-build I don't see any change at http://html5test.com:
```
Video codecs
MPEG-4 ASP support      No ✘
H.264 support           No ✘
H.265 support           No ✘
Ogg Theora support      Yes ✔
WebM with VP8 support   Yes ✔
WebM with VP9 support   Yes ✔
```

iigs commented:
It is my understanding that Firefox (64-bit) looks for /usr/lib/64/libavcodec.so during runtime.  My advice is not to use a build with `CONFIGURE_OPTIONS += --enable-ffmpeg`.
Hipster packages and their version number:
`video/ffmpeg` `2.8.15-2018.0.0.0`
`audio/faad2` `2.7-2018.0.0.1`
`library/video/x264` `0.148.0.20170527-2018.0.0.0`
`library/audio/libmad` 0.15.1.2-2018.0.0.1

truss -f firefox will show when firefox attempts to load libavcodec.so

Google search on the subject of ffmpeg and firefox also suggest to install `library/audio/gstreamer1/plugin/ugly` and `library/audio/gstreamer1/plugin/bad`.

Firefox 60ESR has the capability to support all audio formats except CAF on http://hpr.dogphilosophy.net/test/ (mp3 and aac comes from ffmpeg)

The YouTube HTML5 support page also provides feedback checking for VPx, H.264, and MSE support: https://www.youtube.com/html5

Mno-hime commented:
Thanks @iigs. I added the non-`gstreamer` dependencies you suggested as YouTube was then able to detect H.264, MP3, and AAC.

Overall, the stability is good once I removed uBlock Origin and started off fresh profile.

Biggest hassle is by far non-accelerated video.

iigs commented:
> Thanks @iigs. I added the non-`gstreamer` dependencies you suggested as YouTube was then able to detect H.264, MP3, and AAC.

The depends put into commit 26d8846b9714ba50146e1707a0650d9d83f05c15 are all encumbered.  "hipster" packages should not have any "hipster-encumbered" dependencies.  I think this commit should not be used.

An alternative solution is to have steps on the OI wiki (in a Firefox section) to manually add the encumbered repo and the four packages.

Also, commenting on this line I wrote:
> After using firefox 60.4 for "a while", the firefox process has hung

I have gotten another application to hang on this same system, so this does not appear to be an issue with Firefox.

Mno-hime commented:
Right, I dropped the commit.

pyhalov commented on components/web/firefox/patches/13-rename_log2.patch:
```diff
@@ -1,94 +0,0 @@
-__log2 name conflicts with function in libm,
```
Restore this patch

Mno-hime commented:
With `13-rename_log2.patch` Firefox 60 won't crash when starting with pre-version-60 profile.

Wondering which patches are we missing as well... Needs to be audited.

pyhalov commented on components/web/firefox/Makefile:
```diff
@@ -15,60 +15,81 @@
 COMPONENT_PREP_ACTION = ( cd $(@D) && autoconf-2.13 &&\
                           cd $(@D)/js/src && autoconf-2.13 )
 
 COMPONENT_PRE_CONFIGURE_ACTION = (cp $(COMPONENT_DIR)/files/gld-wrapper  $(BUILD_DIR); chmod +x $(BUILD_DIR)/gld-wrapper )
```
Do we still need gld-wrapper?

pyhalov commented on components/web/firefox/patches/07-branding.patch:
```diff
@@ -1,11 +0,0 @@
---- firefox-52.1.0esr/browser/branding/official/pref/firefox-branding.js.~1~	2017-05-04 11:51:32.593399439 +0300
```
Please, restore this patch in one or another way - we want to display welcome page on first FF run.

pyhalov commented on components/web/firefox/patches/20-sqlite3-unix-excl.patch:
```diff
@@ -1,17 +0,0 @@
-This provides the option to set the indicator for a NFS profile to true.
-It may be necessary for users on NFS to use bookmarks/logging.
-Will not be sending upstream.
```
We likely want this patch in one or another form

Mno-hime commented:
Should `sendfile(3EXT)` still be disabled (https://github.com/OpenIndiana/oi-userland/pull/4656/files#diff-5715b50b7279ba73284543ee81f8a564)?

Mno-hime commented:
@iigs When I move the mouse pointer inside the Firefox window, videos play smoothly, and thing happens with `about:config` page - it filters items only when I move the pointer. Also text cursor in any text area of a web page blinks only when the pointer is moving. Can you see the same?

iigs commented:
Yes.  I also see the identical behavior on Solaris.  If I was to make a guess to explain this, thread(s) that are responsible for drawing on windows are sleeping for too long.

Toasterson commented:
I can confirm that Firefox works except the Video Playback thing. Github now is happy again and Javascript heavy do not slow down every tab anymore. Once the Sleep thing is solved I would say this is good to go.

Mno-hime commented:
I run Firefox on X220 (2 cores/4 threads, 12 GB od RAM) and it for the most part behaves, but when I replaced 6 GB of RAM, Firefox started behaving oddly (stucks for a couple of seconds, dumps core, ...), so I had to decrease "Content process limit" from 4 to 1 to stabilise it, a blurb right next to it even says "Additional content processes can improve performance when using multiple tabs, but will also use more memory.". So, I wonder if we should stick with the default in this regard.

@iigs, @Toasterson How much RAM and cores/threads do you have?

Toasterson commented:
An i7 quad core with Hyperthreading with 32GB RAM. No Problem with it dumping core,

EDIT: Memory overflow Problems maybe? Also for the video?

iigs commented:
Machines I have used 60ESR:
* Core i5 (2/4) 16GB RAM
* Core i7 (4/8) 16GB RAM
* Core i5 (2/4) 8GB RAM

Geoff

iigs commented:
Issues discovered using the site http://codylindley.com/native.html :

1) <input type="color causes"> Firefox crash if the color dropper is used twice. See https://www.illumos.org/issues/10228

2) <input type="range"> does not display a horizontal line delineating the start and stop of range area.  This is a nimbus theme problem.

3) The text area uses a monospace font which defaults to DejaVu Sans Mono 12 point.  On this page, enter nine lines of text.  The size of the text area is user controllable.  Size the form element height so it is less than 4.5 lines of text--the vertical scrollbars do not appear on the widget.  Size the widget height so it is betweem 4.5 and less than 8 lines of text.  The scrollbar slider does not appear in the vertical scrollbar.  The properly functional scrollbar only works if the widget height is 8 or more lines.  Note: the GTK3 rendering of the scrollbar widget is based on pixel size, not line count.  I think this issue is due to limitations in GTK3 deprecating support for scrollbars that do not look/operate like Adwaita.

pyhalov commented:
At least here I see several issues:
1) Scrolling is terribly slow. I see visual artifacts while scrolling, for example, through this PR or switching tabs. 
2) It seems that FF doesn't honor our theme. At least blue border above active tab looks irritating ;)

pyhalov commented:
At least here I see several issues:
1) Scrolling is terribly slow. I see visual artifacts while scrolling, for example, through this PR or switching tabs. 
2) It seems that FF doesn't honor our theme. At least blue border above active tab looks irritating ;)
3) It seems it eats my 8GB RAM much faster now :(

Toasterson commented:
This seems to be theme related. Or amount of tabs related. With one tab open it is using about 800mb.... Welp. But it runs and scrolls smooth with awesome as window manager. But I don't think Linux is any better of with memory consumption. I think every Process has it's own Javascript engine? There could be a huge amount of basic allocation for that. Mybe something to fix for us. Maybe everybody is suffering that.

pyhalov commented:
> At least here I see several issues:
> 
>     1. Scrolling is terribly slow. I see visual artifacts while scrolling, for example, through this PR or switching tabs.

Works fine with Nvidia. Terribly slow on intel/vesa...

Mno-hime commented:
@pyhalov I don't see any artefacts on X220 (Intel GFX). Scrolling is fine for me as well.

Try lowering the amount of threads used, per previous discussion here. (Also there's HW accel disable button).

I am afraid that's how much memory modern web browsers need, may not be related to OI...

pyhalov commented:
http://html5test.com/ reports:
```
MPEG-4 ASP support No 
H.264 support No
H.265 support No 
Ogg Theora support Yes 
WebM with VP8 support Yes
WebM with VP9 support Yes 
```

What did I miss to enable h264?

Toasterson commented:
Probably the ffmpeg stuff that is still not properly due to the 32-64bit issue we had. Does Video Playback work without moving the mouse for you?

pyhalov commented:
Youtube works, VK doesn't. That's a problem for me so far...

Toasterson commented:
Youtube works for you? Even if the mouse is still?

pyhalov commented:
Ouch, outdated 32-bit only ffmpeg...

pyhalov commented:
> Youtube works for you? Even if the mouse is still?

No. It stucks (as well as VK and likely other videos).

RocketMan commented:
I think the sticking HTML5 video in absence of mouse movements may be a symptom of event starvation, problem with the event queue, or interval timer events not being delivered.

Here is something to check:  Open the attachment on this old mozilla slip:  https://bugzilla.mozilla.org/show_bug.cgi?id=907402   The attachment contains a simple JS script where setInterval should be firing continuously.  Expected:  It should continuously print the time to the web console.  For me in 60esr, it prints a few times and stops.  Moving the mouse starts it up again, but only as long as the mouse is in motion, then it stops.  Like the HTML5 issue, for me this is a regression from 52esr.

Unfortunately, the old mozilla slip above and ones related to it do not seem to indicate a root cause or solution.  However, it does suggest the problem goes beyond HTML5 video to event handling.

Toasterson commented:
@RocketMan This is true for me too. 

Are we missing Patches? Has anybody tried on pkgsrc fireofox 60 ?

I see for example https://github.com/joyent/pkgsrc/blob/trunk/www/firefox60/patches/patch-js_src_threading_posix_Thread.cpp

Which might help and we don't have. or also https://github.com/joyent/pkgsrc/blob/trunk/www/firefox60/patches/patch-js_src_util_NativeStack.cpp

iigs commented:
@Toasterson I went ahead and tried those two patches--they make no difference in fixing the thread sticking problem (i.e. playing videos, etc)--I actually see a longer time for Firefox to become useable on first start with these patches (e10s support is enabled).  Illumos added named threads (see committ ab618543cc6fc4bc273c077ef5d247961cdb29d4) on October 19.  I believe those pkgsrc patches are only needed for kernels without named thread support.

Mno-hime commented:
What would be the component affected by "event starvation"? I built Firefox with system libevent instead of the bundled one (the one bundled by Firefox was bumped to v2.1, system one is at v2.0.x), but no change.

RocketMan commented:
The problem with the interval timer and HTML5 playback is this:  The worker thread that manages the timers is being put to sleep for an incorrect amount of time.

The worker thread in the esr codebase uses a `PRIntervalTime` value to communicate the sleep time to `Monitor::Wait`.  `PRIntervalTime` is a platform-dependent value; however `Monitor::Wait` is assuming it is in milliseconds, which is incorrect for Solaris.  `Monitor::Wait` should be calling  `PR_IntervalToMilliseconds` to convert `PRIntervalTime` to ms.

The attached patch resolves the issue for me in 60.4 esr.  Not upstreaming, as the latest FF sources have changed from using the older `PRIntervalTime` to `TimeDuration` for setting the wait time on the thread.

@Toasterson, can you please validate this fix works for you?

Many thanks to @wanderview for his insights into the FF timer code.

[condvar.patch.txt](https://github.com/OpenIndiana/oi-userland/files/2792349/condvar.patch.txt)

pyhalov commented:
Does Adblock Plus extension work for you with FF 60?

iigs commented:
condvar patch fixes these problems:
* video playing
* having a timely response to display URL in the lower left hand corner when hovering over links
* having a timely response with about:config searching
* having a timely response to see the alt text when hovering over an image (displayed like a tool tip)
* setInterval javascript example in mozilla bugid 907402

This was tested on  the Firefox 60.5.0ESR candidate build 1.

iigs commented:
I have noticed this problem with Firefox 60.4 and 60.5 on OpenIndiana (it doesn't happen on Solaris):
During the time when this previously mentioned message appears "[Parent 2002, Gecko_IOThread] WARNING: Failed to launch tab subprocess: file /export/home/newman/ws/oi-userland/components/web/firefox/firefox-60.4.0/ipc/glue/GeckoChildProcessHost.cpp, line 557", and for some time afterwards (a few seconds to a time frame around 20 seconds or so), the memory SIZE as shown in top increases to 7.5GB or more RAM on one of the Firefox processes (e10s enabled and having default number of content processes).  The memory SIZE jumped to 13GB or more on a 16GB RAM machine.  Once the memory SIZE drops down to ~1.5G on that process, then firefox becomes usable.

iigs commented:
The latest Ad Block Plus extension does not work for me.  After initially installing it, the checkboxes in the preferences can not be checked.  Stopping Firefox and starting it, the memory SIZE (based on top) consumes 7.5GB on a 8GB system.  Firefox never becomes usable.  Firefox eventually crashes, taking other desktop apps with it (crashing MATE??).  Maybe some type of fork bomb consuming all stack space?

Toasterson commented:
Yeah there seems to be something going on still. For me the condvar patch does not work at all. Adblock Plus Works for me and I also get the memory Jumping to 8Gb sometimes. Mostly when enabling and disabling Adblock plus. And also during launch of firefox.

RocketMan commented:
I am also finding that with the condvar patch, process shutdown completes right away.  Before the patch, I was finding that process shutdown in 60 esr could run many seconds.

RocketMan commented:
@Toasterson, try `MOZ_LOG=nsTimerImpl:4 firefox`  You will see (among other things) the amount of sleep time requested by the timer thread, the actual sleep time, and the delta between the two.  There is also a summary upon shutdown (mean and stddev).

The timer thread only sleeps in one place, and condvar patches that code path.  Try the setInterval test in mozilla bugid 907402.  With the patch in place, the delta should be no more than about 30ms; mostly, it is around 0ms.

If you are seeing large deltas, it suggests the patch is not present in the build.

iigs commented:
The memory SIZE spike during launch problem and slowness to a usable state seems to be caused by a problem with Web Assembly (wasm).  I didn't check, but that may also explain why Ad Block Plus isn't working.

In about:config , setting javascript.options.wasm to false provides a startup with no issues.  I do not suggest releasing Firefox with wasm disabled.  This information is to help identify the real issue.

Toasterson commented:
@RocketMan Yeah I am getting high deltas. But I also verified that the patch is inside the patches sub directory.
   09-condvar.patch

Do I need to do something special other than pulling the pr locally?

I have a little go wasm programm that I used for testing the WASM stuff.

It's a simple hello world. Already allocating the 7Gb of ram. But not reserving it for the process. I can also add more code to test what is exaclty failing if that helps. like allocation or function calls etc.

edit git repo for the interested in debugging:
https://git.wegmueller.it/Toasterson/webfrontend

Toasterson commented:
@Mno-hime huh what was that? a rebase?

Mno-hime commented:
@Toasterson I am sorry about that, botched `git pull`.

Anyway, also botched `03-fix-MarkPagesInUse-implementation.patch`, Firefox now crashes, I will restore it.

Mno-hime commented:
Botched patch fixed.

In the meantime release 60.5.0 sources are available, will update to it: https://ftp.mozilla.org/pub/firefox/releases/60.5.0esr/source/.

Mno-hime commented:
Turns out hash of the 60.5.0 release is the same as candidate build 2 of 60.5.0.

iigs commented:
Petr from Oracle has been in contact with me about the Firefox porting efforts.  Two things.

1) Has anybody tried to use Firefox with a NFS home directory (or at least the .mozilla directory on NFS)?  I am told that on Solaris, the history isn't stored or retrieved.  Mozilla BugID 1428169 has comments about NFS issues on Linux.  Petr adds that setting storage.nfs_filesystem to true in about:config doesn't help.

2) What I referred to the memory SIZE problem is a swap issue.    swap -sh shows what is allocated, reserved, and available. Per Firefox content process, around 1.5G of swap is used where 1G of it goes to reservation.
  
Before starting Firefox

$ swap -sh
total: 1.5G allocated + 724M reserved = 2.2G used, 6.7G available

After starting Firefox:

$ swap -sh
total: 1.9G allocated + 3.7G reserved = 5.6G used, 3.3G available

It is recommended that at least 6G of swap be available.  Set browser.tabs.remote.autostart to false in about:config to disable multiple content processes.

Mno-hime commented:
@iigs Does WebAssembly work for you? This is what I get with Firefox 60.5.0 on Lenovo X230 (in case GL matters here):

* https://webassembly.org/demo/Tanks/: "The browser could not allocate enough memory for the WebGL content. If you are the developer of this content, try allocating less memory to your WebGL build in the WebGL player settings."
* http://demo.tombo.io/: "Oops! Looks like something went wrong. Please try reloading the page."
* http://s3.amazonaws.com/mozilla-games/ZenGarden/EpicZenGarden.html Won't load.

Mno-hime commented:
I compiled Firefox with `jemalloc` with and without `replace-malloc` and in both cases it dumped core or was otherwise unusable. Though, Solaris has them both enabled: https://github.com/oracle/solaris-userland/blob/master/components/desktop/firefox/Makefile#L105.

Toasterson commented:
@Mno-hime The examples work for me. Although I have a beefy 32Gb Machine. So it is not an infinite allocation. Probably a bit too much though. It probably uses the wrong allocation. SWAP instead of pure memory. Linux and Solaris have a different memory Architecture. Also one game was complaining that I ran a 32-bit firefox. Even though I do run 64-bit acording to the help page.

RocketMan commented:
There is a possible anomaly in rendering a multi-line HTML SELECT element in the nimbus theme.  For example, on this page:  http://zookeeper.stanford.edu/index.php?action=viewDJ&seq=selUser&viewuser=5

Expected: In the list at the bottom of the page, there should be a scrollbar thumb, and clicks in the scroll trough should paginate the list.

iigs commented:
@Mno-hime This is the site I used to test wasm when I claimed it was working:  https://floooh.github.io/oryol/ .  I see now that my swap reservation increases by about 6GB when running a wasm demo.

With the Epic Zen site, I ge this error on the page: 
Building shaders failed:
Shader linking failed: Statically used varyings do not fit within packing limits. (see GLSL ES Specification 1.0.17, p111)
Current renderer: Mozilla Mozilla WebGL 2.0, WebGL GLSL ES 3.00

I see the same errors as you do with Tanks and Tombo.

This is on the computer with 16GB RAM and nVidia 600 series graphics.

Also, not only jemalloc is a difference between Solaris and OI, but so is this patch:
https://github.com/oracle/solaris-userland/blob/master/components/desktop/firefox/patches/Bug577056.patch (specifically the ArrayBufferObject.cpp file at the bottom is not patched on OI).  Adding just this file change causes Firefox to core very early in its execution.

@RocketMan The link with your HTML SELECT comment shows the thumb and scrollbar for me in the widget at the bottom of the page.  I reported the idential issue in my comment https://github.com/OpenIndiana/oi-userland/pull/4656#issuecomment-453787177 .  I incorrectly used the term slider when I meant thumb.

RocketMan commented:
@iigs sorry about that, I did see your comment (it is what inspired me to suspect something similar here) but I did not connect the misbehaviour of TEXTAREA with SELECT.  However, it seems the root cause is the same.

FWIW, for me, webkit correctly renders the widget on the zookeeper page, but checking just now, it fails on the TEXTAREA control on codylindley test page.  So not a Firefox issue; as you suspect, it is likely a GTK3 limitation, or bug in the theme.

Toasterson commented:
This fixed the video Playback for me finally. Is there anything left except the GTK Theme?

Thanks @Mno-hime

Mno-hime commented:
I don't see XPI lang packs in Add-ons (about:addons page) altough we ship them. Does anyone see them in Add-ons page?

iigs commented:
firefox.p5m is configured so an individual language is only installed if facet.locale.xx is true.  This is the method used by Firefox 52ESR on OI. The Solaris version of Firefox 60ESR has all languages installed for all users.  The also means that Solaris Firefox allows for more languages than what pkg's locale check even supports.  Let me know if you would like for me to change Makefile and firefox.p5m to better match what Solaris is doing.

Mno-hime commented:
@iigs Thanks for the explanation! If it's how things worked for Firefox 52 ESR, then be it.

I am for squash and merge then.

Anyone having any concerns?

iigs commented:
I'm not seeing anybody making comments to try to fix the Web Assembly problem where several gigs of RAM gets reserved.  I'm not seeing large amounts of RAM being reserved during Firefox startup on one of my builds.  I have not yet compiled 60.5.1 on my other build system to compare.

If Firefox gets published this way, users will have to be told that if they get a Web Extension that reserves lots of RAM, they may have to use this work-around:
* uninstall the extension (may need to start in safe mode to complete this step)
* install an older version of the extension which does not exhibit the behavior
* configure that extension to not update

Mno-hime commented:
Updated Firefox to 60.5.2.

I saw that solaris-userland adds `$JOBS` in `$COMPONENT_BUILD_ARGS`. Anyone know's about the purpose? I noticed that without it Rust compilation is utilitizing my CPUs to only 40 %. But might by just an illusion on my side.

Added configuration which sets default content processes to two. As OpenIndiana os often run on systems with limited resources, this limits amount of CPU and memory used.

iigs commented:
Follow up from my comment from 10 days ago.  I finally had a chance to look at my second build system.  It had Firefox 60.5.0 installed. Prior to a pkg update, the swap reserved would spike on startup and then drop.  After a pkg update, there is no swap reserve spike. The firefox package did not change.

The swap reserve still increases when accessing a Web page that runs Web Assembly.  I am speculating that perhaps this is a Illumos or userland library problem, and not a problem with Firefox?  Does anybody have any ideas how to troubleshoot this?

The only thing I can confirm is that a BE dated February 18 was when I no longer noticed that the swap reserve spike was happening.

Mno-hime commented:
Squashed commits. I think we are ready for integration.

pyhalov commented:
What about issues with Adblock? Does Solaris Firefox version suffer from it?

pyhalov commented:
Also latest version just dies on html5test.com for me

iigs commented:
There are no issues with Web Assembly with the Solaris Firefox--stack reservation doesn't spike.  Adblock Plus doesn't hang.  The PSPDF webassembly benchmark completes https://pspdfkit.com/webassembly-benchmark/ .  Verified with 60.6ESR.

iigs commented:
I have no problems with html5test.com using 60.6ESR.  I tested with a brand new profile.  I was greeted with 439 out of 555 points. I have no problems with html5test.com using my normal profile as well.

pyhalov commented:
Now it dies even with clean profile shortly after start with
<pre>
[Parent 1566, Gecko_IOThread] WARNING: pipe error: Broken pipe: file /export/home/alp/srcs/oi-userland/components/web/firefox/firefox-60.6.0/ipc/chromium/src/chrome/common/ipc_channel_posix.cc, line 694
</pre> 
in ~/.xsession-errors and 
<pre>
 fffffd7fdea56b5e _ZN7mozilla3ipc14MessageChannel22OnChannelErrorFromLinkEv () + 1ae
 0000000000000000 ???????? ()
</pre>
stack trace.

Mno-hime commented:
Currently build fails for me with:
```
stack backtrace:
   0:           0x82a33e - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::h563f81f6af186079
   1:           0x83eda6 - std::sys_common::backtrace::print::h8ed111738dc46dbe
   2:           0x8334db - std::panicking::default_hook::{{closure}}::h852231eb38896e30
   3:           0x833238 - std::panicking::default_hook::h555d38db9a26f1fd
   4:           0x833bbb - std::panicking::rust_panic_with_hook::h4c5650c84c4200c9
   5:           0x833749 - std::panicking::continue_panic_fmt::h0364dd584f19717c
   6:           0x8336ad - std::panicking::begin_panic_fmt::he8f90578df3611d6
   7:           0x571f34 - build_script_build::build_gecko::bindings::write_binding_file::h09bb52fac816cb29
   8:           0x576b87 - build_script_build::build_gecko::bindings::generate_structs::hcb160ed8c31e9200
   9:           0x84b899 - __rust_maybe_catch_panic
  10:           0x56929b - <F as alloc::boxed::FnBox<A>>::call_box::h9eef45187e7ef3ac
  11:           0x82d9ca - std::sys_common::thread::start_thread::h539e581cbfa146c3
  12:           0x8090a5 - std::sys::unix::thread::Thread::new::thread_start::h13287a420d2408a9
  13: 0xfffffd7feac50179 - _thrp_setup
  14: 0xfffffd7feac5048f - smt_pause
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Any', src/libcore/result.rs:1009:5
stack backtrace:
   0:           0x82a33e - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::h563f81f6af186079
   1:           0x83eda6 - std::sys_common::backtrace::print::h8ed111738dc46dbe
   2:           0x8334db - std::panicking::default_hook::{{closure}}::h852231eb38896e30
   3:           0x833238 - std::panicking::default_hook::h555d38db9a26f1fd
   4:           0x833bbb - std::panicking::rust_panic_with_hook::h4c5650c84c4200c9
   5:           0x833749 - std::panicking::continue_panic_fmt::h0364dd584f19717c
   6:           0x833645 - rust_begin_unwind
   7:           0x86706b - core::panicking::panic_fmt::hdbafadab3061affb
   8:           0x568e8a - core::result::unwrap_failed::hcd6d350af19846bd
   9:           0x57d331 - build_script_build::main::hc7db1b05e5723077
  10:           0x561fe2 - std::rt::lang_start::{{closure}}::h5d982522740644f2
  11:           0x8335e2 - std::panicking::try::do_call::h69db5d46710edf88
  12:           0x84b899 - __rust_maybe_catch_panic
  13:           0x8044e8 - std::rt::lang_start_internal::h8ea26b550d812617
  14:           0x57d850 - main
  15:           0x560af2 - _start_crt
  16:           0x560a57 - _start

make[4]: *** [/export/home/newman/ws/oi-userland/components/web/firefox/firefox-60.6.1/config/rules.mk:979: force-cargo-library-build] Error 101
make[4]: Leaving directory '/export/home/newman/ws/oi-userland/components/web/firefox/build/amd64/toolkit/library/rust'
make[3]: *** [/export/home/newman/ws/oi-userland/components/web/firefox/firefox-60.6.1/config/recurse.mk:73: toolkit/library/rust/target] Error 2
make[3]: *** Waiting for unfinished jobs....
```
It seems @iigs can build it fine.

Mno-hime commented:
> What about issues with Adblock? Does Solaris Firefox version suffer from it?

[Adblock Plus](https://addons.mozilla.org/en-US/firefox/addon/adblock-plus/) works for me.

pyhalov commented:
I have a hypothesis what is going on - we've updated gcc-6, and gcc-6 include paths are hardcoded in clang. FF bindgen uses clang (via bindgen), and clang fails to find GCC headers. Have to rebuild clang.

pyhalov commented:
@Mno-hime  Please, add clang to REQUIRED_PACKAGES

Mno-hime commented:
@iigs has the same hypothesis, his Firefox build fails after GCC 6.5 update.

pyhalov commented:
Installed new FF. So far have the following issues:
- Ad Block Plus still doesn't allow me to change its settings (switched to Ublock Origin), but this is a regression.
- Not all youtube videos play fine (for example, https://www.youtube.com/watch?v=tAGnKpE4NCI several times threw an error). It seems, error is related to Ublock Origin.

Overall, I think we can merge it, at least this will make other people look at issues :)

Mno-hime commented:
>     * Ad Block Plus still doesn't allow me to change its settings (switched to Ublock Origin), but this is a regression.

Worked for me, was able to pick block lists, disable "OK adds" option, ...

>     * Not all youtube videos play fine (for example, https://www.youtube.com/watch?v=tAGnKpE4NCI several times threw an error). It seems, error is related to Ublock Origin.

Just enjoyed it twice (with uBlock installed, Russian block list enable) without any problems.

> Overall, I think we can merge it, at least this will make other people look at issues :)

:).

AndWac commented:
Am 20.04.19 um 13:37 schrieb Michal Nowak:
>
>     |* Ad Block Plus still doesn't allow me to change its settings
>     (switched to Ublock Origin), but this is a regression. |
>
> Worked for me, was able to pick block lists, disable "OK adds" option, ...
>
>     |* Not all youtube videos play fine (for example,
>     https://www.youtube.com/watch?v=tAGnKpE4NCI several times threw an
>     error). It seems, error is related to Ublock Origin. |
>
> Just enjoyed it twice (with uBlock installed, Russian block list
> enable) without any problems.
>
>     Overall, I think we can merge it, at least this will make other
>     people look at issues :)
>
> :).
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenIndiana/oi-userland/pull/4656#issuecomment-485099216>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADET4NYZ3SLSPUDNSOZ6WU3PRL56ZANCNFSM4GK3GJNQ>.
>
Maybe it's possible to add it additionally to the old ESR version, at
least until the new version gets stable enough?

Mno-hime commented:
> Maybe it's possible to add it additionally to the old ESR version, at least until the new version gets stable enough?

Using 60 ESR for several months, I think it's stable enough, for sure it's not perfect, but it's good enough. Having old ESR around means less people using the new one, devs being less inclined towards fixing the new one. Also many sites does not work with old ESR, tha't making OI looking bad for newcommers.

Mno-hime commented:
I removed `21-limit-content-processes.patch` which was setting default context processes to 1. The reasoning behind this patch was it makes Firefox use less memory and thus making it more stable, however the actuall problem was [me being low on SWAP](https://echelog.com/logs/browse/oi-dev/1555711200). Extending it to be 1:1 with physical memory, makes Firefox not being killed when other memory hungry apps are running (e.g. Thunderbird, `pkg`).

AndWac commented:
Your reasoning is correct but nevertheless it should be possible to have
an alternative for those having problems with the new one.
So, make the new one the default and it case of trouble allow
installation of the old one.



Am 20.04.19 um 14:24 schrieb Michal Nowak:
>
>     Maybe it's possible to add it additionally to the old ESR version,
>     at least until the new version gets stable enough?
>
> Using 60 ESR for several months, I think it's stable enough, for sure
> it's not perfect, but it's good enough. Having old ESR around means
> less people using the new one, devs being less inclined towards fixing
> the new one. Also many sites does not work with old ESR, tha't making
> OI looking bad for newcommers.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenIndiana/oi-userland/pull/4656#issuecomment-485108374>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADET4N4MXL5CGOBRNBF2LP3PRMDRBANCNFSM4GK3GJNQ>.
>

Mno-hime commented:
> Your reasoning is correct but nevertheless it should be possible to have an alternative for those having problems with the new one. So, make the new one the default and it case of trouble allow installation of the old one.

OK, we can always do this, if there are problems we can't fix in the new ESR. I wouldn't oppose that.

Mno-hime commented:
Slightly off topic, but Firefox 68 ESR is comming in June and Firefox 60 ESR is EOL in October:

![](https://www.mozilla.org/media/img/firefox/organizations/release-overview-high-res.d3c9788f2dd1.png)

(Taked from https://www.mozilla.org/en-US/firefox/organizations.)

So, yes, it's a good time to work on current non-ESR Firefox stream, so we are ready for Firefox 68 ESR.
